rust/src/mime/smtp.rs

```diff
@@ -0,0 +1,825 @@
+use md5::Md5;
+use std::ffi::CStr;
+use std::io;
+use std::os::raw::c_uchar;
+
+#[repr(u8)]
+#[derive(Copy, Clone, Debug, PartialOrd, PartialEq, Eq)]
+pub enum MimeSmtpParserState {
+    MimeSmtpStart = 0,
+    MimeSmtpHeader = 1,
+    MimeSmtpBody = 2,
+    MimeSmtpParserError = 3,
+}
+
+impl Default for MimeSmtpParserState {
+    fn default() -> Self {
+        MimeSmtpParserState::MimeSmtpStart
+    }
+}
+
+#[derive(Debug, Default)]
+pub struct MimeHeader {
+    pub name: Vec<u8>,
+    pub value: Vec<u8>,
+}
+
+#[repr(u8)]
+#[derive(Copy, Clone, Debug, PartialOrd, PartialEq, Eq)]
+pub enum MimeSmtpMd5State {
+    MimeSmtpMd5Disabled = 0,
+    MimeSmtpMd5Inited = 1,
+    MimeSmtpMd5Started = 2,
+    MimeSmtpMd5Completed = 3,
+}
+
+#[repr(u8)]
+#[derive(Copy, Clone, Debug, PartialOrd, PartialEq, Eq)]
+enum MimeSmtpContentType {
+    Message = 0,
+    PlainText = 1,
+    Html = 2,
+    Unknown = 3,
+}
+
+impl Default for MimeSmtpContentType {
+    fn default() -> Self {
+        MimeSmtpContentType::Message
+    }
+}
+
+#[derive(Debug)]
+pub struct MimeStateSMTP<'a> {
+    pub(crate) state_flag: MimeSmtpParserState,
+    pub(crate) headers: Vec<MimeHeader>,
+    pub(crate) main_headers_nb: usize,
+    filename: Vec<u8>,
+    pub(crate) attachments: Vec<Vec<u8>>,
+    pub(crate) urls: Vec<Vec<u8>>,
+    boundary: Vec<u8>,
+    encoding: MimeSmtpEncoding,
+    decoder: Option<MimeBase64Decoder>,
+    content_type: MimeSmtpContentType,
+    decoded_line: Vec<u8>,
+    // small buffer for end of line
+    // waiting to see if it is part of the boundary
+    bufeol: [u8; 2],
+    bufeolen: u8,
+    files: &'a mut FileContainer,
+    sbcfg: *const StreamingBufferConfig,
+    md5: md5::Md5,
+    pub(crate) md5_state: MimeSmtpMd5State,
+    pub(crate) md5_result: GenericArray<u8, U16>,
+}
+
+#[derive(Debug)]
+pub struct MimeBase64Decoder {
+    tmp: [u8; 4],
+    nb: u8,
+}
+
+impl MimeBase64Decoder {
+    pub fn new() -> MimeBase64Decoder {
+        MimeBase64Decoder { tmp: [0; 4], nb: 0 }
+    }
+}
+
+impl Default for MimeBase64Decoder {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+pub fn mime_smtp_state_init(
+    files: &mut FileContainer, sbcfg: *const StreamingBufferConfig,
+) -> Option<MimeStateSMTP> {
+    let r = MimeStateSMTP {
+        state_flag: MimeSmtpParserState::MimeSmtpStart,
+        headers: Vec::new(),
+        main_headers_nb: 0,
+        filename: Vec::new(),
+        attachments: Vec::new(),
+        urls: Vec::new(),
+        boundary: Vec::new(),
+        decoded_line: Vec::new(),
+        encoding: MimeSmtpEncoding::Plain,
+        decoder: None,
+        content_type: MimeSmtpContentType::Message,
+        bufeol: [0; 2],
+        bufeolen: 0,
+        files,
+        sbcfg,
+        md5: Md5::new(),
+        md5_state: MimeSmtpMd5State::MimeSmtpMd5Disabled,
+        md5_result: [0; 16].into(),
+    };
+    return Some(r);
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn rs_mime_smtp_state_init(
```

Comment:
Public C API so should follow those naming conventions prefixed with `SC`.  See `rust/src/ffi/hashing.rs` as an example.

Reply:
For all `git grep 'pub unsafe extern "C"' rust/src/mime/` ?

Only for new ones ?
(I see that http2 for instance has the same `rs_http_parse_content_range`)

Reply:
Commit pushed in my branch multipart-mime-3487-v23 ;-)